File: results/0.0/ARP/arp-claude-3-5-sonnet-20241022.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    uint16_t hardware_type;
    uint16_t protocol_type;
    uint8_t hardware_addr_len;
    uint8_t protocol_addr_len;
    uint16_t operation;
    uint8_t* sender_hardware_addr;
    uint8_t* sender_protocol_addr;
    uint8_t* target_hardware_addr;
    uint8_t* target_protocol_addr;
} arp_packet;

static HParser* init_arp_parser(void) {
    HParser* hardware_type = h_uint16();
    HParser* protocol_type = h_uint16();
    HParser* hardware_addr_len = h_uint8();
    HParser* protocol_addr_len = h_uint8();
    HParser* operation = h_uint16();
    
    // Variable length fields based on length fields
    HParser* sender_hardware_addr = h_length_value(hardware_addr_len, h_uint8());
    HParser* sender_protocol_addr = h_length_value(protocol_addr_len, h_uint8());
    HParser* target_hardware_addr = h_length_value(hardware_addr_len, h_uint8());
    HParser* target_protocol_addr = h_length_value(protocol_addr_len, h_uint8());

    return h_sequence(hardware_type, protocol_type, hardware_addr_len, 
                     protocol_addr_len, operation, sender_hardware_addr,
                     sender_protocol_addr, target_hardware_addr, 
                     target_protocol_addr, NULL);
}

static void print_bytes(uint8_t* data, size_t len) {
    for(size_t i = 0; i < len; i++) {
        printf("%02x", data[i]);
    }
    printf("\n");
}

static void print_arp_packet(const HParsedToken* token) {
    if(token->token_type != TT_SEQUENCE) return;
    
    printf("Hardware Type: 0x%04x\n", (uint16_t)token->seq->elements[0]->uint);
    printf("Protocol Type: 0x%04x\n", (uint16_t)token->seq->elements[1]->uint);
    printf("Hardware Address Length: %u\n", (uint8_t)token->seq->elements[2]->uint);
    printf("Protocol Address Length: %u\n", (uint8_t)token->seq->elements[3]->uint);
    printf("Operation: %u\n", (uint16_t)token->seq->elements[4]->uint);
    
    printf("Sender Hardware Address: ");
    print_bytes(token->seq->elements[5]->bytes.token, token->seq->elements[5]->bytes.len);
    
    printf("Sender Protocol Address: ");
    print_bytes(token->seq->elements[6]->bytes.token, token->seq->elements[6]->bytes.len);
    
    printf("Target Hardware Address: ");
    print_bytes(token->seq->elements[7]->bytes.token, token->seq->elements[7]->bytes.len);
    
    printf("Target Protocol Address: ");
    print_bytes(token->seq->elements[8]->bytes.token, token->seq->elements[8]->bytes.len);
}

int main(int argc, char *argv[]) {
    if(argc != 2) {
        fprintf(stderr, "Usage: %s <arp_packet_file>\n", argv[0]);
        return 1;
    }

    FILE *f = fopen(argv[1], "rb");
    if(!f) {
        perror("Failed to open file");
        return 1;
    }

    fseek(f, 0, SEEK_END);
    size_t size = ftell(f);
    fseek(f, 0, SEEK_SET);

    uint8_t *input = malloc(size);
    if(!input) {
        perror("Failed to allocate memory");
        fclose(f);
        return 1;
    }

    if(fread(input, 1, size, f) != size) {
        perror("Failed to read file");
        free(input);
        fclose(f);
        return 1;
    }

    HParser *parser = init_arp_parser();
    const HParsedToken *token = h_parse(parser, input, size);

    if(token) {
        print_arp_packet(token);
        h_parse_result_free(token);
    } else {
        fprintf(stderr, "Failed to parse ARP packet\n");
    }

    free(input);
    fclose(f);
    return 0;
}

File: results/0.0/ARP/arp-claude-3-5-sonnet-20241022.rs
use nom::{
    bytes::complete::take,
    combinator::map,
    number::complete::{be_u16, be_u8},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
pub struct ArpPacket {
    hardware_type: u16,
    protocol_type: u16,
    hardware_addr_len: u8,
    protocol_addr_len: u8,
    operation: u16,
    sender_hardware_addr: Vec<u8>,
    sender_protocol_addr: Vec<u8>,
    target_hardware_addr: Vec<u8>,
    target_protocol_addr: Vec<u8>,
}

fn parse_arp_packet(input: &[u8]) -> IResult<&[u8], ArpPacket> {
    let (input, hardware_type) = be_u16(input)?;
    let (input, protocol_type) = be_u16(input)?;
    let (input, hardware_addr_len) = be_u8(input)?;
    let (input, protocol_addr_len) = be_u8(input)?;
    let (input, operation) = be_u16(input)?;
    
    let (input, sender_hardware_addr) = map(
        take(hardware_addr_len),
        |bytes: &[u8]| bytes.to_vec()
    )(input)?;
    
    let (input, sender_protocol_addr) = map(
        take(protocol_addr_len),
        |bytes: &[u8]| bytes.to_vec()
    )(input)?;
    
    let (input, target_hardware_addr) = map(
        take(hardware_addr_len),
        |bytes: &[u8]| bytes.to_vec()
    )(input)?;
    
    let (input, target_protocol_addr) = map(
        take(protocol_addr_len),
        |bytes: &[u8]| bytes.to_vec()
    )(input)?;

    Ok((input, ArpPacket {
        hardware_type,
        protocol_type,
        hardware_addr_len,
        protocol_addr_len,
        operation,
        sender_hardware_addr,
        sender_protocol_addr,
        target_hardware_addr,
        target_protocol_addr,
    }))
}

impl ArpPacket {
    fn display_addresses(&self) {
        println!("Hardware Type: {:#04x}", self.hardware_type);
        println!("Protocol Type: {:#04x}", self.protocol_type);
        println!("Hardware Address Length: {}", self.hardware_addr_len);
        println!("Protocol Address Length: {}", self.protocol_addr_len);
        println!("Operation: {}", self.operation);
        
        print!("Sender Hardware Address: ");
        for byte in &self.sender_hardware_addr {
            print!("{:02x}:", byte);
        }
        println!();
        
        print!("Sender Protocol Address: ");
        for byte in &self.sender_protocol_addr {
            print!("{}.", byte);
        }
        println!();
        
        print!("Target Hardware Address: ");
        for byte in &self.target_hardware_addr {
            print!("{:02x}:", byte);
        }
        println!();
        
        print!("Target Protocol Address: ");
        for byte in &self.target_protocol_addr {
            print!("{}.", byte);
        }
        println!();
    }
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <arp_packet_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_arp_packet(&buffer) {
        Ok((remaining, packet)) => {
            if !remaining.is_empty() {
                println!("Warning: {} bytes of trailing data", remaining.len());
            }
            packet.display_addresses();
        }
        Err(e) => {
            eprintln!("Failed to parse ARP packet: {:?}", e);
            std::process::exit(1);
        }
    }

    Ok(())
}